DM-44147: Allow alternate dimensions in file template

timj opened the pull request:
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

timj committed:
Allow alternates in file template

This is mainly to allow day_obs/group to be used directly in
templates with newer dimension universes whilst also supporting
older universes.

Now you can say '{exposure|group}' and only use group in the
template if there is no exposure in the dataId. Similarly for
day_obs you can say '{day_obs|visit.day_obs|exposure.day_obs}'
and pick the first relevant one. The alternates can be removed
when we are sure no older universe registries exist.

timj committed:
Use alternates in file template and merge some options

{exposure.obs_id}_{group} -> {exposure.obs_id|group}
{visit.day_obs}/{exposure.day_obs} -> {day_obs|visit.day_obs|exposure.day_obs}

timj committed:
Add news fragment

timj commented on python/lsst/daf/butler/datastore/file_templates.py:
```diff
@@ -522,6 +531,24 @@ def format(self, ref: DatasetRef) -> str:
+                alternates = field_name.split("|")
+                for alt in alternates:
+                    if "." in alt:
+                        primary, _ = alt.split(".")
```
In theory we could extend the test here so that if someone asked for `exposure.x|exposure` (say, because `x` was new in the universe but if it wasn't there fall back to `exposure`) fairly simply. I'm not sure if it's worth doing though.

TallJimbo replied:
If we don't have a use case let's not do it.  Even though I buy that the approach you've taken on this ticket is the minimal change to solve the problem at hand, I'd still like to search in the future for ways to simplify all of this by offering less functionality, as opposed to extending that functionality further.

TallJimbo commented on python/lsst/daf/butler/datastore/file_templates.py:
```diff
@@ -406,25 +414,26 @@ def grouped_fields(self) -> tuple[dict[str, set[str]], dict[str, set[str]]]:
             "parent": set(),
         }
 
```
Preexisting, but it'd be great to annotate these return objects with `TypedDict` (or just use dataclasses if that wouldn't break anything).

timj replied:
I used `TypedDict` -- it did require a slight reorganization of the code since you can't say `mydict[var]` -- it has to be a literal. I didn't use a dataclass because that would be a much bigger change and this method is a bit of a hotspot.

TallJimbo commented on python/lsst/daf/butler/datastore/file_templates.py:
```diff
@@ -406,25 +414,26 @@ def grouped_fields(self) -> tuple[dict[str, set[str]], dict[str, set[str]]]:
+                        # This needs to be added twice.
+                        subfield = field_name
+                        key = "parent"
+                        field_name, _ = field_name.split(".")
```
I think the logic would be clearer if you moved the `if "?"` block that sets `target` above this one, so you don't need the separate `if subfield` later.

timj committed:
Reorganize logic in template fields to save an if statement

timj committed:
Use TypedDict for grouped_fields

This required a minor reorganization since TypedDict mandates
literal string fields.
